File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        
        unordered_map<int,int> ump;
        int i = 0,j = 0;
        int ans = 0;
        int n = fruits.size();
        while(j<n) {
            ump[fruits[j]]++;
            if(ump.size()>2) {
                ump[fruits[i]]--;
                if(ump[fruits[i]]==0) {
                    ump.erase(fruits[i]);
                }
                i++;
            }
            if(ump.size()<=2) {
                ans = max(ans,j-i+1);
            }
            j++;
        }

        return ans;
    }
};

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        Map<Integer,Integer> mp = new HashMap<>();
        int n = fruits.length;
        int  i = 0;
        int j = 0;
        int ans = 0;
        while(j<n) {
            
            if(mp.size()==2 && mp.getOrDefault(fruits[j],-1)==-1) {
                ans = Math.max(ans,j-i);
                int key = -1;
                int kv = Integer.MAX_VALUE;
                for(int k: mp.keySet()) {
                    if(mp.get(k)<=kv) {
                        kv = mp.get(k);
                        key = k;
                    } 
                }
                i = Math.max(i,mp.get(key)+1);
              
                mp.remove(key);
            }
            
            mp.put(fruits[j],j);

            j++;
        }
        ans = Math.max(ans,j-i);
        
        return ans;
    }
}